### PR TITLE
Mark enums with non-zero discriminant as non-zero

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rfcs/issues/1230
r? @eddyb

### CODE REVIEW

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -769,6 +770,7 @@ pub enum Layout {
     CEnum {
         discr: Integer,
         signed: bool,
+        non_zero: bool,
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: You already have the range (see `rustc_trans::adt` for a way to determine if a value is in the range), why would this additional field be needed?

File: src/librustc/ty/layout.rs [hunk 2]
```diff
@@ -1076,6 +1080,12 @@ impl<'a, 'gcx, 'tcx> Layout {
                                         value: value
                                     });
                                 }
+                                CEnum { discr: value, .. } => {
+                                    return success(RawNullablePointer {
+                                        nndiscr: discr as u64,
+                                        value: Int(value)
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Maybe make this `match` produce the `Primitive` type and do the return after the `match`?

File: src/tools/compiletest/src/main.rs [hunk 3]
```diff
@@ -15,6 +15,8 @@
 #![feature(test)]
 #![feature(libc)]
 
+#![cfg_attr(stage0, feature(question_mark))]
```
Comment (eddyb) [inline src/tools/compiletest/src/main.rs]: A few people hit this, I guess it got removed by accident. Thanks!

**Review by eddyb — CHANGES_REQUESTED** [review]
LGTM although I believe it could be simpler.

File: src/tools/compiletest/src/main.rs (hunk 3 above)
Comment (petrochenkov) [inline src/tools/compiletest/src/main.rs]: https://github.com/rust-lang/rust/pull/37202 contains it too :D
The sooner it lands the better.

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (petrochenkov) [inline src/librustc/ty/layout.rs]:
```
enum E {
    A = -1,
    B = 1,
}
```
`0` is in the range, but the enum is still non-zero.
(Also, `non_zero: bool` fits into previous padding so it doesn't increase the size of `Layout`)

Comment (eddyb) [inline src/librustc/ty/layout.rs]: That's annoying. What I expect to eventually have would use `2` there, that's why I don't like `non_zero`: it's a special-case that makes sense mostly on pointers.
If you want to discuss the full niche optimization, I'd be down for it (IRC?).
If you don't want to change this implementation I'll r+ and take out this later (and/or switch to multi-ranges - LLVM actually supports them for `!range`).

**Comment by Manishearth** [comment]
Btw, this caused some rather tricky issues in Servo (https://github.com/servo/rust-bindgen/issues/225).

IMO when changing reprs for things which may come from FFI (e.g. an enum) we should be more careful and have a cycle of warnings or a compiler note or something. Not sure how it would be implemented. (I agree that it's non ideal for C++ to abuse enums as bitflags but apparently it's a common pattern)

The only reason we could figure this out was because I knew how Rust's encoded enum debuginfo works, and I also knew that this PR had happened -- which is a rather high bar for being able to debug things like this. It's otherwise quite an opaque and nonsensical bug that feels like a compiler bug (in this case, setting `foo = Some(thing)` made it stay `None`)

cc @eddyb @nikomatsakis

**Comment by nikomatsakis** [comment]
Hmm. @Manishearth I've not read this PR, but I assume that the enums in question were not marked as `#[repr(C)]` or any such thing?

**Comment by Manishearth** [comment]
They were marked as `repr(u8)` which IIRC is equivalent?

**Comment by eddyb** [comment]
That's just setting the size of the enum, it's still UB to have a value other than the assigned ones.

**Pull request merged**